tensorflow/core/kernels/data/dataset_test_base.h

```diff
@@ -0,0 +1,191 @@
+limitations under the License.
+==============================================================================*/
+
+#ifndef TENSORFLOW_CORE_KERNELS_DATA_DATASET_TEST_BASE_H
+#define TENSORFLOW_CORE_KERNELS_DATA_DATASET_TEST_BASE_H
+
+#include <memory>
+#include <tuple>
+#include <vector>
+
+#include "tensorflow/core/framework/dataset.h"
+#include "tensorflow/core/framework/function.h"
+#include "tensorflow/core/framework/function_handle_cache.h"
+#include "tensorflow/core/framework/function_testlib.h"
+#include "tensorflow/core/framework/node_def_builder.h"
+#include "tensorflow/core/framework/partial_tensor_shape.h"
+#include "tensorflow/core/framework/variant.h"
+#include "tensorflow/core/framework/variant_tensor_data.h"
+#include "tensorflow/core/kernels/data/dataset_utils.h"
+#include "tensorflow/core/kernels/data/iterator_ops.h"
+#include "tensorflow/core/kernels/ops_testutil.h"
+#include "tensorflow/core/platform/test.h"
+#include "tensorflow/core/platform/types.h"
+#include "tensorflow/core/util/ptr_util.h"
+
+namespace tensorflow {
+namespace data {
+
+// Helpful functions to test dataset kernels.
+class DatasetOpsTestBase : public ::testing::Test {
+ public:
+  DatasetOpsTestBase()
+      : device_(DeviceFactory::NewDevice("CPU", {}, "/job:a/replica:0/task:0")),
+        device_type_(DEVICE_CPU) {
+    CHECK(device_.get()) << "Could not create CPU device";
+    allocator_ = device_->GetAllocator(AllocatorAttributes());
+  }
+
+  ~DatasetOpsTestBase() {
+    gtl::STLDeleteElements(&tensors_);
+    gtl::STLDeleteElements(&managed_outputs_);
+  }
+
+  // Creates a new operation based on the node definition.
+  std::unique_ptr<OpKernel> CreateOpKernel(const NodeDef& node_def);
+
+  // Creates a new dataset. Here we assume that the dataset operation has only
+  // one output stored in the OpKernelContext.
+  DatasetBase* CreateDataset(OpKernel* kernel, OpKernelContext* context);
+
+  // Creates a new iterator for iterating over the range of elements in this
+  // dataset. Meanwhile, `iterator_context` will be initialized internally.
+  //
+  // This method may be called multiple times on the same dataset, and the
+  // resulting iterators will have distinct state.
+  std::unique_ptr<IteratorBase> CreateIterator(
+      OpKernelContext* context, DatasetBase* dataset,
+      std::unique_ptr<IteratorContext>* iterator_context);
+
+  // Gets the next output from the range that this iterator is traversing.
+  //
+  // If at least one output remains in this iterator's range, that
+  // output will be stored in `*out_tensors` and `false` will be
+  // stored in `*end_of_sequence`.
+  //
+  // If no more outputs remain in this iterator's range, `true` will
+  // be stored in `*end_of_sequence`, and the content of
+  // `*out_tensors` will be undefined.
+  Status GetNext(IteratorBase* iterator, IteratorContext* iterator_context,
+                 std::vector<Tensor>* out_tensors, bool* end_of_sequence);
+
+  // Creates a new range dataset operation.
```

Comment:
1) change docstring to `Creates a new RangeDataset op kernel`
2) change signature to propagate status to the caller
3) document the template type